docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -346,58 +384,75 @@ spec:
+    - Good for scheduling upgrades/SLOs
+    - Set `upgradeAfter` to now (in RFC3339 form) if an upgrade is required immediately
+
+- The controller should tolerate the manual or automatic removal of a replica during the upgrade process. A replica that fails during the upgrade may block the completion of the upgrade. Removal or other remedial action may be necessary to allow the upgrade to complete.
+
+- In order to determine if a Machine to be rolled out, KCP implements the following:
+    - The infrastructureRef link used by each machine at creation time is stored in annotations at machine level.
+    - The kubeadmConfigSpec used by each machine at creation time is stored in annotations at machine level.
+        - If the annotation is not present (machine is either old or adopted), we won't roll out on any possible changes made in KCP's ClusterConfiguration given that we don't have enough information to make a decision. Users should use KCP.Spec.UpgradeAfter field to force a rollout in this case.
+
+##### RolloutType CreateFirst (create-and-delete)
+
+CreateFirst is the default rollout strategy of `KubeadmControlPlane.Spec.RolloutStrategy` or user can set it to "CreateFirst" to enable following rollout algorithm:  
+
   - Find Machines that have an outdated spec
   - If there is a machine requiring rollout
     - Scale up control plane creating a machine with the new spec
     - Scale down control plane by removing one of the machine that needs rollout (the oldest out-of date machine in the failure domain that has the most control-plane machines on it)
-    
-- In order to determine if a Machine to be rolled out, KCP implements the following:
-    - The infrastructureRef link used by each machine at creation time is stored in annotations at machine level.
-    - The kubeadmConfigSpec used by each machine at creation time is stored in annotations at machine level.
-        - If the annotation is not present (machine is either old or adopted), we won't roll out on any possible changes made in KCP's ClusterConfiguration given that we don't have enough information to make a decision.
-           Users should use KCP.Spec.UpgradeAfter field to force a rollout in this case.
-    
-- The controller should tolerate the manual or automatic removal of a replica during the upgrade process. A replica that fails during the upgrade may block the completion of the upgrade. Removal or other remedial action may be necessary to allow the upgrade to complete.
+
+##### RolloutType DeleteFirst (delete-and-create)
+
+DeleteFirst rollout strategy is performed when `KubeadmControlPlane.Spec.RolloutStrategy` is set to "DeleteFirst". DeleteFirst rollout strategy is recommended for resource constrained environment like bare-metal, OpenStack or vSphere resource pools, etc.
+
+- The algorithm is the following:
+  - Webhook validation verifies that control plane replica count is >= 3
+    - If replica count is less than 3 user is informed that rollout can not be performed
```

Comment:
I'm wondering if it would be better to fallback to `CreateFirst` in the event that it's not safe to delete an instance for purposes of maintaining etcd quorum rather than trying to block operations through webhook validation?